applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp

```diff
@@ -213,28 +212,50 @@ Vector& ElasticIsotropic3D::CalculateValue(
     if (rThisVariable == STRAIN ||
         rThisVariable == GREEN_LAGRANGE_STRAIN_VECTOR ||
         rThisVariable == ALMANSI_STRAIN_VECTOR) {
-        this->CalculateCauchyGreenStrain( rParameterValues, rValue);
+
+        Flags& r_flags = rParameterValues.GetOptions();
+
+        // Previous flags saved
+        const bool flag_const_tensor = r_flags.Is( ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR );
+        const bool flag_stress = r_flags.Is( ConstitutiveLaw::COMPUTE_STRESS );
+
+        // Set flags to only compute the strain
+        r_flags.Set(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR, false);
+        r_flags.Set(ConstitutiveLaw::COMPUTE_STRESS, false);
+
+        ElasticIsotropic3D::CalculateMaterialResponsePK2(rParameterValues);
+        rValue = rParameterValues.GetStrainVector();
+
+        // Previous flags restored
+        r_flags.Set( ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR, flag_const_tensor );
+        r_flags.Set( ConstitutiveLaw::COMPUTE_STRESS, flag_stress );
+
     } else if (rThisVariable == STRESSES ||
         rThisVariable == CAUCHY_STRESS_VECTOR ||
         rThisVariable == KIRCHHOFF_STRESS_VECTOR ||
         rThisVariable == PK2_STRESS_VECTOR) {
-        // Get Values to compute the constitutive law:
+
         Flags& r_flags = rParameterValues.GetOptions();
 
         // Previous flags saved
         const bool flag_const_tensor = r_flags.Is( ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR );
         const bool flag_stress = r_flags.Is( ConstitutiveLaw::COMPUTE_STRESS );
 
+        // Set flags to only compute the stress
         r_flags.Set( ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR, true );
         r_flags.Set( ConstitutiveLaw::COMPUTE_STRESS, true );
 
-        // We compute the stress
-        ElasticIsotropic3D::CalculateMaterialResponseCauchy(rParameterValues);
+        ElasticIsotropic3D::CalculateMaterialResponsePK2(rParameterValues);
         rValue = rParameterValues.GetStressVector();
 
         // Previous flags restored
         r_flags.Set( ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR, flag_const_tensor );
         r_flags.Set( ConstitutiveLaw::COMPUTE_STRESS, flag_stress );
+
+    } else if (rThisVariable == INITIAL_STRAIN_VECTOR) {
+        if (this->HasInitialState()) {
+           rValue = GetInitialState().GetInitialStrainVector();
+        }
```

Comment:
Maybe assign an empty Vector here?
Otherwise the output might be messed up I think

Reply:
I am confused too

Reply:
You mean to set rValue to empty before the "if", right?

Reply:
no an `else`
~~~c++
        if (this->HasInitialState()) {
           rValue = GetInitialState().GetInitialStrainVector();
        } else {
          rValue = ZeroVector(0);
        }
~~~

I am not sure though bcs of the usage

what do you think?

Reply:
Makes sense. Done.

Reply:
what about the noalias?